stdlib/runtime/numberUtils.gr

```diff
@@ -0,0 +1,1334 @@
+    for (let mut i = length; WasmI32.ltS(i, kk); i += 1n) {
+      WasmI32.store8(buffer + i, _CHAR_CODE_0, 0n)
+    }
+    WasmI32.store16(buffer + kk, _CHAR_CODE_DOT | (_CHAR_CODE_0 << 8n), 0n)
+    kk + 2n
+  } else if (WasmI32.gtS(kk, 0n) && WasmI32.leS(kk, 21n)) {
+    // 1234e-2 -> 12.34
+    let ptr = buffer + kk
+    Memory.copy(
+      ptr + 1n,
+      ptr,
+      0n - k
+    )
+    WasmI32.store8(buffer + kk, _CHAR_CODE_DOT, 0n)
+    length + 1n
+  } else if (WasmI32.ltS(-6n, kk) && WasmI32.leS(kk, 0n)) {
+    // 1234e-6 -> 0.001234
+    let offset = 2n - kk
+    Memory.copy(
+      buffer + offset,
+      buffer,
+      length
+    )
+    WasmI32.store16(buffer, _CHAR_CODE_0 | (_CHAR_CODE_DOT << 8n), 0n)
+    for (let mut i = 2n; WasmI32.ltS(i, offset); i += 1n) {
+      WasmI32.store8(buffer + i, _CHAR_CODE_0, 0n)
+    }
+    length + offset
+  } else if (length == 1n) {
+    // 1e30
+    WasmI32.store8(buffer, _CHAR_CODE_e, 1n)
+    length = genExponent(buffer + 2n, kk - 1n)
+    length + 2n
+  } else {
+    let len = length
+    Memory.copy(
+      buffer + 2n,
+      buffer + 1n,
+      len - 1n
+    )
+    WasmI32.store8(buffer,       _CHAR_CODE_DOT, 1n)
+    WasmI32.store8(buffer + len, _CHAR_CODE_e,   1n)
+    length += genExponent(buffer + len + 2n, kk - 1n)
+    length + 2n
+  }
+}
+
+let dtoa_core = (buffer, value) => {
+  let mut value = value
+  let hasSign = WasmF64.lt(value, 0.W)
+  if (hasSign) {
+    value = WasmF64.neg(value)
+    WasmI32.store8(buffer, _CHAR_CODE_MINUS, 0n)
+  }
+  let sign = if (hasSign) 1n else 0n
+  let len = grisu2(value, buffer, sign)
+  let len = prettify(buffer + sign, len - sign, _K)
+  len + sign
+}
+
+let mut _dtoa_buf = -1n
+
+let get_dtoa_buf = () => {
+  if (_dtoa_buf == -1n) {
+    _dtoa_buf = Memory.malloc(_MAX_DOUBLE_LENGTH);
+  }
+  _dtoa_buf
+}
+
+let isFinite = (value) => {
+  WasmF64.eq(WasmF64.sub(value, value), 0.W)
+}
+
+let isNaN = (value) => {
+  WasmF64.ne(value, value)
+}
+
+export let dtoa = (value) => {
+  let str = if (WasmF64.eq(value, 0.W)) {
+    let ret = allocateString(3n)
+    WasmI32.store8(ret, _CHAR_CODE_0, 8n)
+    WasmI32.store8(ret, _CHAR_CODE_DOT, 9n)
+    WasmI32.store8(ret, _CHAR_CODE_0, 10n)
+    ret
+  } else if (!isFinite(value)) {
+    if (isNaN(value)) {
+      let ret = allocateString(3n)
+      WasmI32.store(ret, 0x4E614En, 8n) // NaN
+      ret
+    } else if (WasmF64.lt(value, 0.W)) {
+      let ret = allocateString(9n)
+      WasmI64.store(ret, 0x74696E69666E492DN, 8n) // tinifnI-
+      WasmI32.store8(ret, 0x79n, 16n) // y
+      ret
+    } else {
+      let ret = allocateString(8n)
+      WasmI64.store(ret, 0x7974696E69666E49N, 8n) // ytinifnI
+      ret
+    }
+  } else {
+    let size = dtoa_core(get_dtoa_buf(), value)
+    let result = allocateString(size)
+    Memory.copy(result + 8n, get_dtoa_buf(), size)
+    result
+  }
+  WasmI32.toGrain(str): String
+}
+
+
+export let dtoa_buffered = (buffer: usize, value: f64) => {
```

Comment:
@ospencer said that we don't use this right now, so should we open an issue to consider removing it and only add it back when needed? Heck, even just toss the code into the issue.

Reply:
https://github.com/grain-lang/grain/issues/555